README


This archive contains the distribution Config-INI,
version 0.024:

  simple .ini-file format

This software is copyright (c) 2007 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.015.



Changes
Changelog for Config-INI

0.024     2014-04-05 21:31:54-04:00 America/New_York
        - MIGHT BREAK YOUR CODE (but probably won't):
          handle_unparsed_line now receives its arguments in reversed order
        - new methods exist and more methods now receive the filehandle to make
          subclassing more flexible

0.023     2014-03-13 10:21:15+01:00 Europe/Paris
        - specify minimum Mixin::Linewise requirement (thanks, Andreas Koenig
          and Smylers)

0.022     2014-01-30 16:57:43-05:00 America/New_York
        - remove the last few places IO::String was used (thanks, Graham Knop)

0.021     2013-10-20 11:33:59 America/New_York
        - throw an exception if line 1 appears to start with a UTF-8 BOM
        - typo fixes in docs (thanks, David Steinbrunner!)

0.020     2013-07-01 21:30:50 America/New_York
          update bugtracker

0.019     2011-12-15 17:03:14 America/New_York
          provide handle_unparsed_line

0.018     2011-06-02 21:51:50 America/New_York
          bump up version of IO::File required (thanks, ABH!)

0.017     2010-12-11 10:20:34 America/New_York
          avoid test failures on Win32 by using . instead of TMPDIR
          for writing test files (thanks, Ahmad M. Zawawi)

0.016     2010-09-04
          add Mixin::Linewise to CARP_NOT -- better reporting of error location

0.015     2010-08-22
          documentation improvements (thanks, castaway)

0.014     2009-01-16
          add repo to metadata

0.013     2008-06-05
          refactored some of the read_* write_* stuff to Mixin::Linewise

0.012     2008-04-12
          BUG FIX: allow for different error message on cygwin

          fixed typo in docs (thanks HDP)

0.011     2007-11-19
          BUG FIX: "0" was not a valid section name

0.010     2007-11-04
          BUG FIX: detect when we're dumping something we won't be able to read
          BUG FIX: fix tests when filetest.pm is not available (pre-5.6)
            (thanks to CPAN tester srezic)
          DOCS: documentation of core file format in Config::INI

0.009     2007-11-01
          add placeholding Config::INI (content soon!)

0.008     2007-10-31
          update Module::Install

0.007     2007-08-21
          use filetest.pm in tests to avoid -r-related failures on Cygwin

0.006     2007-08-15
          use -T instead of -t in tests; -t is a perl 5.008 feature

0.005     2007-08-13
          add finalize method, lazy-load starting section method (this means
          that you can have starting_section die to prevent properties outside
          of explicity setting)

          added Config::INI::Writer (largely done by Florian Ragwitz)

          further improved subclassability of Reader (thanks again, Florian!)

0.004     2007-05-12
          fix asinine error in test, poorly adapted from Config::Tiny

0.003     2007-05-10
          first release


lib/Config/INI.pm
use strict;
use warnings;
package Config::INI;
# ABSTRACT: simple .ini-file format
$Config::INI::VERSION = '0.023';
# =head1 SYNOPSIS
#
# Config-INI comes with code for reading F<.ini> files:
#
#   my $config_hash = Config::INI::Reader->read_file('config.ini');
#
# ...and for writing C<.ini> files:
#
#   Config::INI::Writer->write_file({ somekey => 'somevalue' }, 'config.ini');
#
# See L<Config::INI::Writer> and L<Config::INI::Reader> for more examples.
#
# =head1 GRAMMAR
#
# This section describes the format parsed and produced by Config::INI::Reader
# and ::Writer.  It is not an exhaustive and rigorously tested formal grammar,
# it's just a description of this particular implementation of the
# not-quite-standardized "INI" format.
#
#   ini-file   = { <section> | <empty-line> }
#
#   empty-line = [ <space> ] <line-ending>
#
#   section        = <section-header> { <value-assignment> | <empty-line> }
#
#   section-header = [ <space> ] "[" <section-name> "]" [ <space> ] <line-ending>
#   section-name   = string
#
#   value-assignment = [ <space> ] <property-name> [ <space> ]
#                      "="
#                      [ <space> ] <value> [ <space> ]
#                      <line-ending>
#   property-name    = string-without-equals
#   value            = string
#
#   comment     = <space> ";" [ <string> ]
#   line-ending = [ <comment> ] <EOL>
#
#   space = ( <TAB> | " " ) *
#   string-without-equals = string - "="
#   string = ? 1+ characters; not ";" or EOL; begins and ends with non-space ?
#
# Of special note is the fact that I<no> escaping mechanism is defined, meaning
# that there is no way to include an EOL or semicolon (for example) in a value,
# property name, or section name.  If you need this, either subclass, wait for a
# subclass to be written for you, or find one of the many other INI-style parsers
# on the CPAN.
#
# The order of sections and value assignments within a section are not
# significant, except that given multiple assignments to one property name within
# a section, only the final one is used.  A section name may be used more than
# once; this will have the identical meaning as having all property assignments
# in all sections of that name in sequence.
#
# =head1 DON'T FORGET
#
# The definitions above refer to the format used by the Reader and Writer classes
# bundled in the Config-INI distribution.  These classes are designed for easy
# subclassing, so it should be easy to replace their behavior with whatever
# behavior your want.
#
# Patches, feature requests, and bug reports are welcome -- but I'm more
# interested in making sure you can write a subclass that does what you need, and
# less in making Config-INI do what you want directly.
#
# =head1 THANKS
#
# Thanks to Florian Ragwitz for improving the subclassability of Config-INI's
# modules, and for helping me do some of my first merging with git(7).
#
# =head1 ORIGIN
#
# Originaly derived from L<Config::Tiny>, by Adam Kennedy.
#
# =cut

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Config::INI - simple .ini-file format

=head1 VERSION

version 0.023

=head1 SYNOPSIS

Config-INI comes with code for reading F<.ini> files:

  my $config_hash = Config::INI::Reader->read_file('config.ini');

...and for writing C<.ini> files:

  Config::INI::Writer->write_file({ somekey => 'somevalue' }, 'config.ini');

See L<Config::INI::Writer> and L<Config::INI::Reader> for more examples.

=head1 GRAMMAR

This section describes the format parsed and produced by Config::INI::Reader
and ::Writer.  It is not an exhaustive and rigorously tested formal grammar,
it's just a description of this particular implementation of the
not-quite-standardized "INI" format.

  ini-file   = { <section> | <empty-line> }

  empty-line = [ <space> ] <line-ending>

  section        = <section-header> { <value-assignment> | <empty-line> }

  section-header = [ <space> ] "[" <section-name> "]" [ <space> ] <line-ending>
  section-name   = string

  value-assignment = [ <space> ] <property-name> [ <space> ]
                     "="
                     [ <space> ] <value> [ <space> ]
                     <line-ending>
  property-name    = string-without-equals
  value            = string

  comment     = <space> ";" [ <string> ]
  line-ending = [ <comment> ] <EOL>

  space = ( <TAB> | " " ) *
  string-without-equals = string - "="
  string = ? 1+ characters; not ";" or EOL; begins and ends with non-space ?

Of special note is the fact that I<no> escaping mechanism is defined, meaning
that there is no way to include an EOL or semicolon (for example) in a value,
property name, or section name.  If you need this, either subclass, wait for a
subclass to be written for you, or find one of the many other INI-style parsers
on the CPAN.

The order of sections and value assignments within a section are not
significant, except that given multiple assignments to one property name within
a section, only the final one is used.  A section name may be used more than
once; this will have the identical meaning as having all property assignments
in all sections of that name in sequence.

=head1 DON'T FORGET

The definitions above refer to the format used by the Reader and Writer classes
bundled in the Config-INI distribution.  These classes are designed for easy
subclassing, so it should be easy to replace their behavior with whatever
behavior your want.

Patches, feature requests, and bug reports are welcome -- but I'm more
interested in making sure you can write a subclass that does what you need, and
less in making Config-INI do what you want directly.

=head1 THANKS

Thanks to Florian Ragwitz for improving the subclassability of Config-INI's
modules, and for helping me do some of my first merging with git(7).

=head1 ORIGIN

Originaly derived from L<Config::Tiny>, by Adam Kennedy.

=head1 AUTHOR

Ricardo Signes <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Config/INI/Reader.pm
use strict;
use warnings;
package Config::INI::Reader;
$Config::INI::Reader::VERSION = '0.023';
use Mixin::Linewise::Readers 0.100;
# ABSTRACT: a subclassable .ini-file parser

# =head1 SYNOPSIS
#
# If F<family.ini> contains:
#
#   admin = rjbs
#
#   [rjbs]
#   awesome = yes
#   height = 5' 10"
#
#   [mj]
#   awesome = totally
#   height = 23"
#
# Then when your program contains:
#
#   my $hash = Config::INI::Reader->read_file('family.ini');
#
# C<$hash> will contain:
#
#   {
#     '_'  => { admin => 'rjbs' },
#     rjbs => {
#       awesome => 'yes',
#       height  => q{5' 10"},
#     },
#     mj   => {
#       awesome => 'totally',
#       height  => '23"',
#     },
#   }
#
# =head1 DESCRIPTION
#
# Config::INI::Reader is I<yet another> config module implementing I<yet another>
# slightly different take on the undeniably easy to read L<".ini" file
# format|Config::INI>.  Its default behavior is quite similar to that of
# L<Config::Tiny>, on which it is based.
#
# The chief difference is that Config::INI::Reader is designed to be subclassed
# to allow for side-effects and self-reconfiguration to occur during the course
# of reading its input.
#
# =cut

use Carp ();

our @CARP_NOT = qw(Mixin::Linewise::Readers);

# =head1 METHODS FOR READING CONFIG
#
# These methods are all that most users will need: they read configuration from a
# source of input, then they return the data extracted from that input.  There
# are three reader methods, C<read_string>, C<read_file>, and C<read_handle>.
# The first two are implemented in terms of the third.  It iterates over lines in
# a file, calling methods on the reader when events occur.  Those events are
# detailed below in the L</METHODS FOR SUBCLASSING> section.
#
# All of the reader methods return an unblessed reference to a hash.
#
# All throw an exception when they encounter an error.
#
# =head2 read_file
#
#   my $hash_ref = Config::INI::Reader->read_file($filename);
#
# Given a filename, this method returns a hashref of the contents of that file.
#
# =head2 read_string
#
#   my $hash_ref = Config::INI::Reader->read_string($string);
#
# Given a string, this method returns a hashref of the contents of that string.
#
# =head2 read_handle
#
#   my $hash_ref = Config::INI::Reader->read_handle($io_handle);
#
# Given an IO::Handle, this method returns a hashref of the contents of that
# handle.
#
# =cut

sub read_handle {
  my ($invocant, $handle) = @_;

  my $self = ref $invocant ? $invocant : $invocant->new;

  # parse the file
  LINE: while (my $line = $handle->getline) {
    if ($handle->input_line_number == 1 && $line =~ /\A\x{FEFF}/) {
      Carp::confess("input handle appears to start with a BOM");
    }

    next LINE if $self->can_ignore($line);

    $self->preprocess_line(\$line);

    # Handle section headers
    if (defined (my $name = $self->parse_section_header($line))) {
      # Create the sub-hash if it doesn't exist.
      # Without this sections without keys will not
      # appear at all in the completed struct.
      $self->change_section($name);
      next LINE;
    }

    if (my ($name, $value) = $self->parse_value_assignment($line)) {
      $self->set_value($name, $value);
      next LINE;
    }

    $self->handle_unparsed_line($handle, $line);
  }

  $self->finalize;

  return $self->{data};
}

# =head1 METHODS FOR SUBCLASSING
#
# These are the methods you need to understand and possibly change when
# subclassing Config::INI::Reader to handle a different format of input.
#
# =head2 current_section
#
#   my $section_name = $reader->current_section;
#
# This method returns the name of the current section.  If no section has yet
# been set, it returns the result of calling the C<starting_section> method.
#
# =cut

sub current_section {
  defined $_[0]->{section} ? $_[0]->{section} : $_[0]->starting_section;
}

# =head2 parse_section_header
#
#   my $name = $reader->parse_section_header($line);
#
# Given a line of input, this method decides whether the line is a section-change
# declaration.  If it is, it returns the name of the section to which to change.
# If the line is not a section-change, the method returns false.
#
# =cut

sub parse_section_header {
  return $1 if $_[1] =~ /^\s*\[\s*(.+?)\s*\]\s*$/;
  return;
}

# =head2 change_section
#
#   $reader->change_section($section_name);
#
# This method is called whenever a section change occurs in the file.
#
# The default implementation is to change the current section into which data is
# being read and to initialize that section to an empty hashref.
#
# =cut

sub change_section {
  my ($self, $section) = @_;

  $self->{section} = $section;

  if (!exists $self->{data}{$section}) {
    $self->{data}{$section} = {};
  }
}

# =head2 parse_value_assignment
#
#   my ($name, $value) = $reader->parse_value_assignment($line);
#
# Given a line of input, this method decides whether the line is a property
# value assignment.  If it is, it returns the name of the property and the value
# being assigned to it.  If the line is not a property assignment, the method
# returns false.
#
# =cut

sub parse_value_assignment {
  return ($1, $2) if $_[1] =~ /^\s*([^=\s\pC][^=\pC]*?)\s*=\s*(.*?)\s*$/;
  return;
}

# =head2 set_value
#
#   $reader->set_value($name, $value);
#
# This method is called whenever an assignment occurs in the file.  The default
# behavior is to change the value of the named property to the given value.
#
# =cut

sub set_value {
  my ($self, $name, $value) = @_;

  $self->{data}{ $self->current_section }{$name} = $value;
}

# =head2 starting_section
#
#   my $section = Config::INI::Reader->starting_section;
#
# This method returns the name of the starting section.  The default is: C<_>
#
# =cut

sub starting_section { q{_} }

# =head2 can_ignore
#
#   do_nothing if $reader->can_ignore($line)
#
# This method returns true if the given line of input is safe to ignore.  The
# default implementation ignores lines that contain only whitespace or comments.
#
# =cut

sub can_ignore {
  my ($self, $line) = @_;

  # Skip comments and empty lines
  return $line =~ /\A\s*(?:;|$)/ ? 1 : 0;
}

# =head2 preprocess_line
#
#   $reader->preprocess_line(\$line);
#
# This method is called to preprocess each line after it's read but before it's
# parsed.  The default implementation just strips inline comments.  Alterations
# to the line are made in place.
#
# =cut

sub preprocess_line {
  my ($self, $line) = @_;

  # Remove inline comments
  ${$line} =~ s/\s+;.*$//g;
}

# =head2 handle_unparsed_line
#
#   $reader->handle_unparsed_line( $io, $line );
#
# This method is called when the reader encounters a line that doesn't look like
# anything it recognizes.  By default, it throws an exception.
#
# =cut

sub handle_unparsed_line {
  my ($self, $handle, $line) = @_;
  my $lineno = $handle->input_line_number;
  Carp::croak "Syntax error at line $lineno: '$line'";
}

# =head2 finalize
#
#   $reader->finalize;
#
# This method is called when the reader has finished reading in every line of the
# file.
#
# =cut

sub finalize { }

# =head2 new
#
#   my $reader = Config::INI::Reader->new;
#
# This method returns a new reader.  This generally does not need to be called by
# anything but the various C<read_*> methods, which create a reader object only
# ephemerally.
#
# =cut

sub new {
  my ($class) = @_;

  my $self = { data => {}, };

  bless $self => $class;
}

# =head1 ORIGIN
#
# Originaly derived from L<Config::Tiny>, by Adam Kennedy.
#
# =cut

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Config::INI::Reader - a subclassable .ini-file parser

=head1 VERSION

version 0.023

=head1 SYNOPSIS

If F<family.ini> contains:

  admin = rjbs

  [rjbs]
  awesome = yes
  height = 5' 10"

  [mj]
  awesome = totally
  height = 23"

Then when your program contains:

  my $hash = Config::INI::Reader->read_file('family.ini');

C<$hash> will contain:

  {
    '_'  => { admin => 'rjbs' },
    rjbs => {
      awesome => 'yes',
      height  => q{5' 10"},
    },
    mj   => {
      awesome => 'totally',
      height  => '23"',
    },
  }

=head1 DESCRIPTION

Config::INI::Reader is I<yet another> config module implementing I<yet another>
slightly different take on the undeniably easy to read L<".ini" file
format|Config::INI>.  Its default behavior is quite similar to that of
L<Config::Tiny>, on which it is based.

The chief difference is that Config::INI::Reader is designed to be subclassed
to allow for side-effects and self-reconfiguration to occur during the course
of reading its input.

=head1 METHODS FOR READING CONFIG

These methods are all that most users will need: they read configuration from a
source of input, then they return the data extracted from that input.  There
are three reader methods, C<read_string>, C<read_file>, and C<read_handle>.
The first two are implemented in terms of the third.  It iterates over lines in
a file, calling methods on the reader when events occur.  Those events are
detailed below in the L</METHODS FOR SUBCLASSING> section.

All of the reader methods return an unblessed reference to a hash.

All throw an exception when they encounter an error.

=head2 read_file

  my $hash_ref = Config::INI::Reader->read_file($filename);

Given a filename, this method returns a hashref of the contents of that file.

=head2 read_string

  my $hash_ref = Config::INI::Reader->read_string($string);

Given a string, this method returns a hashref of the contents of that string.

=head2 read_handle

  my $hash_ref = Config::INI::Reader->read_handle($io_handle);

Given an IO::Handle, this method returns a hashref of the contents of that
handle.

=head1 METHODS FOR SUBCLASSING

These are the methods you need to understand and possibly change when
subclassing Config::INI::Reader to handle a different format of input.

=head2 current_section

  my $section_name = $reader->current_section;

This method returns the name of the current section.  If no section has yet
been set, it returns the result of calling the C<starting_section> method.

=head2 parse_section_header

  my $name = $reader->parse_section_header($line);

Given a line of input, this method decides whether the line is a section-change
declaration.  If it is, it returns the name of the section to which to change.
If the line is not a section-change, the method returns false.

=head2 change_section

  $reader->change_section($section_name);

This method is called whenever a section change occurs in the file.

The default implementation is to change the current section into which data is
being read and to initialize that section to an empty hashref.

=head2 parse_value_assignment

  my ($name, $value) = $reader->parse_value_assignment($line);

Given a line of input, this method decides whether the line is a property
value assignment.  If it is, it returns the name of the property and the value
being assigned to it.  If the line is not a property assignment, the method
returns false.

=head2 set_value

  $reader->set_value($name, $value);

This method is called whenever an assignment occurs in the file.  The default
behavior is to change the value of the named property to the given value.

=head2 starting_section

  my $section = Config::INI::Reader->starting_section;

This method returns the name of the starting section.  The default is: C<_>

=head2 can_ignore

  do_nothing if $reader->can_ignore($line)

This method returns true if the given line of input is safe to ignore.  The
default implementation ignores lines that contain only whitespace or comments.

=head2 preprocess_line

  $reader->preprocess_line(\$line);

This method is called to preprocess each line after it's read but before it's
parsed.  The default implementation just strips inline comments.  Alterations
to the line are made in place.

=head2 handle_unparsed_line

  $reader->handle_unparsed_line( $io, $line );

This method is called when the reader encounters a line that doesn't look like
anything it recognizes.  By default, it throws an exception.

=head2 finalize

  $reader->finalize;

This method is called when the reader has finished reading in every line of the
file.

=head2 new

  my $reader = Config::INI::Reader->new;

This method returns a new reader.  This generally does not need to be called by
anything but the various C<read_*> methods, which create a reader object only
ephemerally.

=head1 ORIGIN

Originaly derived from L<Config::Tiny>, by Adam Kennedy.

=head1 AUTHOR

Ricardo Signes <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "simple .ini-file format",
  "AUTHOR" => "Ricardo Signes <rjbs\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Config-INI",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Config::INI",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Mixin::Linewise::Readers" => "0.100",
    "Mixin::Linewise::Writers" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "IO::File" => 0,
    "List::Util" => 0,
    "Test::More" => "0.96",
    "version" => 0
  },
  "VERSION" => "0.024",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "IO::File" => 0,
  "List::Util" => 0,
  "Mixin::Linewise::Readers" => "0.100",
  "Mixin::Linewise::Writers" => 0,
  "Test::More" => "0.96",
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



